### PR TITLE
Fix job and config identification by matching with URL.

### PR DESCRIPTION
**Pull request by mst-ableton**
This builds up a list of jobs and activeConfigurations from the first
Jenkins API 'jobs' call and associates a root URL with each job and
config. Later on, tasks and jobs can match their URL against this list
to identify their job and configuration without any kind of funky
parsing.

### CODE REVIEW

File: jenkins_exporter.py [hunk 1]
```diff
@@ -103,48 +116,32 @@ def _request_queue(self, jobs):
...
             old_max = max_queue_time[job_name][config_name]
             max_queue_time[job_name][config_name] = max(old_max, queuing_time)
 
-        for job, configs in job_names.iteritems():
-            for config in configs:
-                self._prom_metrics['queue'].add_metric(
-                    [job, config], max_queue_time[job][config])
-                self._prom_metrics['queue_count'].add_metric(
-                    [job, config], task_count[job][config])
+        for (job, config) in config_mapping.itervalues():
```
Comment (ala-ableton) [inline jenkins_exporter.py]: :robot: Style nitpicking, the parens around `job, config` are not necessary.

File: jenkins_exporter.py [hunk 2]
```diff
@@ -103,48 +116,32 @@ def _request_queue(self, jobs):
...
-                            job_name = upstream_proj
+            task_url = task.get('task', {}).get('url', '')
+            job_name, config_name = (
+                self.null_configuration_name, self.null_configuration_name)
+            # Look for the longest matching URL in the config mapping.
+            match_length = 0
+            for url, match in config_mapping.iteritems():
+                if task_url.startswith(url) and len(url) > match_length:
+                    job_name, config_name = match
+                    match_length = len(url)
```
Comment (ala-ableton) [inline jenkins_exporter.py]: @mst-ableton what is `match_length` used for?

File: jenkins_exporter.py [hunk 3]
```diff
@@ -81,11 +85,20 @@ def parsejobs(myurl):
...
             return jobs
 
-        return parsejobs(url)
+        jobs = parsejobs(url)
+        config_mapping = {}
+        for job in jobs:
+            if 'activeConfigurations' in job:
+                for config in job['activeConfigurations']:
+                    config_mapping[config['url']] = (job['name'], config['name'])
+            config_mapping[job['url']] = (
```
Comment (ala-ableton) [inline jenkins_exporter.py]: @mst-ableton isn't there an `else` missing before that statement?

File: jenkins_exporter.py [hunk 4]
```diff
@@ -27,11 +27,14 @@ def __init__(self, target, user, password):
         if user and password:
             self._auth = (user, password)
 
+        # Use '__none__' to signify jobs without multiple configurations.
+        self.null_configuration_name = '__none__'
```
Comment (ala-ableton) [inline jenkins_exporter.py]: @mst-ableton this should be a global constant.

File: jenkins_exporter.py (hunk 2 above)
Comment (mst-ableton) [reply jenkins_exporter.py]: line 127:  "Look for the longest matching URL in the config mapping."

if we have a job URL without a config, and a job URL *with* a config, we want the longer one (with the config)_ to be selected

File: jenkins_exporter.py (hunk 3 above)
Comment (mst-ableton) [reply jenkins_exporter.py]: no, we want to add the raw job URL without the configuration as well. before jobs launch, they stay pending with no configuration and we want those to be in the mapping as well

Comment (mst-ableton) [reply jenkins_exporter.py]: ah, I guess I can strip out the 'if' as well then

Comment (ala-ableton) [reply jenkins_exporter.py]: You still need to accommodate for the key `activeConfigurations` not being present. However, I would write `for config in job.get('activeConfigurations', []):`.

Comment (mst-ableton) [reply jenkins_exporter.py]: yep!

File: jenkins_exporter.py [hunk 5]
```diff
@@ -122,8 +122,7 @@ def _request_queue(self, jobs, config_mapping):
 
         for task in result['items']:
             task_url = task.get('task', {}).get('url', '')
-            job_name, config_name = (
-                self.null_configuration_name, self.null_configuration_name)
+            job_name, config_name = (NULL_CONFIGURATION_NAME, NULL_CONFIGURATION_NAME)
```
Comment (ala-ableton) [inline jenkins_exporter.py]: @mst-ableton why are we assigning `NULL_CONFIGURATION_NAME` to a job name?

File: jenkins_exporter.py [hunk 6]
```diff
@@ -89,8 +89,7 @@ def parsejobs(myurl):
         jobs = parsejobs(url)
         config_mapping = {}
         for job in jobs:
-            if 'activeConfigurations' in job:
-                for config in job['activeConfigurations']:
+                for config in job.get('activeConfigurations', []):
                     config_mapping[config['url']] = (job['name'], config['name'])
```
Comment (ala-ableton) [inline jenkins_exporter.py]: The indentation is wacky.

File: jenkins_exporter.py (hunk 5 above)
Comment (mst-ableton) [reply jenkins_exporter.py]: I've rewritten the logic to detect the (hopefully rare) condition where a job is added between the jobs scan (which builds the mapping) and the queue scan, resolving this issue.

**Comment by ala-ableton** [comment]
Once rebased, LGTM :+1:

**Pull request merged**